### ISSUE TITLE
replicating action recognition accuracy for EPIC-KITCHENS-100

### ISSUE THREAD

**Issue by iranroman** [facebookresearch/omnivore#37]
Hello Omnivore!

Thank you very much for everything you have done and provided through this repository. 

I'm interested in building-upon your work for egocentric action recognition, and my first step was to replicate the action recognition results quoted in your paper for the EK100 dataset. If I understand correctly, using the provided `omnivore_swimB_epic` checkpoint, I should be able to obtain an action recognition accuracy close to 50% on the validation subset of the data. However, trying, I only got an accuracy of 35.91% (for top1 action). 

This is a similar problem as the one listed in #20 . However, that issue was closed without a clear resolution. I hope you can help me find what I might be missing. Here are more details about what I am doing: 

### Loading the data

I adapted the epic kitchens data loader found [here](https://github.com/epic-kitchens/epic-kitchens-slowfast/blob/master/slowfast/datasets/epickitchens.py) to do the following:

For each action listed in the [Epic-Kitchens validation csv](https://github.com/epic-kitchens/epic-kitchens-100-annotations/blob/master/EPIC_100_validation.csv):

* ~~find the center frame~~
* ~~load 32 sequential frames (separated with a stride of 2), starting at the center frame. The resulting shape is `[1, 3, 256, 456]`~~
* ~~divide all loaded frame values by `255`~~
* ~~subtract the omnivore `mean = [0.485, 0.456, 0.406] from each channel~~
* ~~divide by the omnivore `std =  [0.229, 0.224, 0.225]`~~
* ~~crop the center `[256x256]` pixels in the image to yield the shape `[1,3,256,256]`~~

I'm tracking my work in this repository: https://github.com/iranroman/ego_actrecog_analysis

Any hints will be greatly appreciated. Thank you very much in advance!

**Comment by Vincent6896** [comment]
Hi. May I ask whether you can load the pretrained checkpoint downloaded in the github instead of use torch.hub.load() function? Thank you very much.

**Comment by iranroman** [comment]
Hi! Thank you very much for your reply. 

I'm loading the model like it is done in the Colab tutorial:

`model = torch.hub.load("facebookresearch/omnivore:main", model=model_name)`

Moreover, I went ahead and manually downloaded the checkpoint from the [`link`](https://dl.fbaipublicfiles.com/omnivore/models/swinB_epic_checkpoint.torch) in the omnivore `README`. I loaded the checkpoint (there were some issues with the key names during the `loading_state_dict` call, but I fixed them this way):

```
model_name = "omnivore_swinB_epic"
model = torch.hub.load("facebookresearch/omnivore:main", model=model_name)
model = model.to(device)
checkpoint = torch.load(cfg.MODEL.CHECKPOINT) # <- here cfg.MODEL.CHECKPOINT = 'swinB_epic_checkpoint.torch', which is the downloadable checkpoint on the github repo
checkpoint_fixed = {} 
for k,v in checkpoint['trunk'].items():
    checkpoint_fixed[f'trunk.{k}']=v
for k,v in checkpoint['heads'].items():
    checkpoint_fixed[f'heads.{k}']=v  
model.load_state_dict(checkpoint_fixed)
```
Running inference with this model gave me the same result. 

Finally, I checked if the values of `checkpoint_fixed` and `model` (before `model.load_state_dict(checkpoint_fixed)`) are the same, which turned out to be true for all the parameters in the model. 

In conclusion, yes, I am using the checkpoint that is downloadable from the github repo (and it is the same as the one provided by `torch.hub.load`). 

Thank you very much for your guidance.

**Comment by Vincent6896** [comment]
Thank you so much for your kind reply. May I ask do you also use the model to extract the feature of the input video/image? If so, how to remove the last classification layers and get the features? Thank you very much.

**Comment by rohitgirdhar** [comment]
Hi @iranroman 
Thanks for your interest in this work.
As mentioned in #20 our data loading hyperparameters for EK are similar to SSv2 in Video Swin Transformer (https://github.com/SwinTransformer/Video-Swin-Transformer/blob/master/configs/recognition/swin/swin_base_patch244_window1677_sthv2.py).
Hence, in the transform pipeline, we use:
1. Temporal segment based sampling -- the whole video is divided into 32 segments, and the center frame is picked from each. For inference this would simply mean uniformly sampling 32 frames (and not at a stride of 2)
2. We then scale the frames such that the short side is 224px
3. Normalize by mean [0.485, 0.456, 0.406] and std [0.229, 0.224, 0.225]
4. Then we spatially crop 3 clips of 224x224 (left, center and right)
The final prediction is the mean of the predictions on all 3 clips.

You can also look at the "Inference with Egocentric videos" section in https://github.com/facebookresearch/omnivore/blob/main/inference_tutorial.ipynb

**Comment by iranroman** [comment]
Thank you @rohitgirdhar 

I have followed your instructions, and I have been able to get an accuracy of 44.65%. Not quite the 47% quoted on the paper, but close enough? 

My code is here in case you are curious. Maybe I'm still doing something wrong:
https://github.com/iranroman/ego_actrecog_analysis

I will look more closely to make sure I'm using the correct data loading hyperparameters.

**Comment by rohitgirdhar** [comment]
That is strange. Perhaps one way to debug could be to try the original VideoSwin-B checkpoint (https://github.com/SwinTransformer/Video-Swin-Transformer) with this code, and see if you can repro the EK or SSv2 numbers reported in our table 6. If those also don't match, then it might indicate something is still off in the preprocessing.

**Comment by iranroman** [comment]
Thank you @rohitgirdhar 

Have you considered sharing the code you use to run the entire evaluation, including the data loader? That would definitively let me know if there's something off. Cheers!

**Comment by iranroman** [comment]
Hi @rohitgirdhar, here's an update:

I revised my data processing approach using the pipeline you pointed me to and I was able to bring the top1 action accuracy up to 45.79% using the provided pre-trained model on the epic kitchens 100 validation set. 

This time I'm 100% sure that I'm correctly processing the data. I even saved post-processed examples as image files to visually inspect if I'm properly resizing and cropping. Everything checks out. 

It is also worth noting that I also integrated the `SlowFast` model in the same codebase, and with that model I was able to perfectly replicate their results on the epic kitchens validation set. 

Here's the repo with my code if you are curious. 
https://github.com/iranroman/ego_actrecog_analysis

45.79% is still pretty good though. It's just not the 49.9% that is expected.

**Comment by rohitgirdhar** [comment]
Hi @iranroman  thanks for the update.
I'm checking with my colleagues who trained/released the model to try to figure what might be causing this discrepancy. Will get back.

**Comment by iranroman** [comment]
Final update. 

I was able to replicate the quoted Omnivore numbers of Epic-Kitchens 100. The remaining issue was that I was using `BGR` instead of `RGB`. Everything worked as expected after addressing that. 

The repository https://github.com/iranroman/ego_actrecog_analysis has the required code and instructions to replicate my observation.

Thank you @rohitgirdhar for your help. I'm closing this issue.

**Comment by Hou9612** [comment]
> Final update.
> 
> I was able to replicate the quoted Omnivore numbers of Epic-Kitchens 100. The remaining issue was that I was using `BGR` instead of `RGB`. Everything worked as expected after addressing that.
> 
> The repository https://github.com/iranroman/ego_actrecog_analysis has the required code and instructions to replicate my observation.
> 
> Thank you @rohitgirdhar for your help. I'm closing this issue.

Hello, thanks very much for your useful discussion in this issue, I'm also interested in the egocentric action recognition task. Could please share your code?

**Comment by iranroman** [comment]
@Hou9612 the code is public

**Comment by Hou9612** [comment]
> @Hou9612 the code is public

I have cloned the code. Thanks very much！
